File: src/element.cpp
/* taverner

By theadventureofe(John Gormley)

element source code

the_adventure_of_e λ */

//some of these may be removable at a later stage
#include <iostream>
#include <utility>
#include <string>
#include <vector>

#include "../headers/element.hpp"

Element::Element()
{
    this->words =
    {
        {"names", vec_str{"ELEMENT_NAME_DEFAULT"}},
        {"adjectives", vec_str{"ELEMENT_ADJECTIVES_DEFAULT"}}
    };
}

Element::Element(element_words words)
{
    this->words =
    {
        {"names", words.names},
        {"adjectives", words.adjectives}
    };
}

void Element::print()
{
    std::cout << "(ELEMENT-PRINT)\n";
    std::cout << "(NAME): " << this->words["names"][0] << "\n";
    std::cout << "\n";
}

vec_str element_set_names(std::initializer_list<std::string> names)
{
    vec_str result;

    for(std::string i : names)
    {
      result.push_back(i);
    }

    return result;
}

vec_str element_set_adjectives(std::initializer_list<std::string> adjectives)
{
    vec_str result;

    for(std::string i : adjectives)
    {
      result.push_back(i);
    }

    return result;
}

std::string Element::name()
{
  return this->words[0][0];
}

Element alien = Element
(
    element_words
    {
      .names = element_set_names({"alien", "extraterrestrial"}),
      .adjectives = element_set_adjectives({"unearthly", "otherworldy"}) 
    }
);

Element air = Element
(
    element_words
    {
      .names = element_set_names({"air", "wind"}),
      .adjectives = element_set_adjectives({"windy"}) 
    }
);

Element chaos = Element
(
    element_words
    {
      .names = element_set_names({"chaos"}),
      .adjectives = element_set_adjectives({"chaotic"}) 
    }
);

Element earth  = Element
(
    element_words
    {
      .names = element_set_names({"earth", "soil", "mud", "dirt"}),
      .adjectives = element_set_adjectives({"earthy", "muddy"}) 
    }
);

Element electricity  = Element
(
    element_words
    { 
      .names = element_set_names({"electricity", "shock", "lightning"}), 
      .adjectives = element_set_adjectives({"electric", "electricifying", "shocking"}) 
    }
);

Element evil  = Element
(
    element_words
    {
      .names = element_set_names({"evil", "wickedness"}),
      .adjectives = element_set_adjectives({"unholy", "wicked"}) 
    }
);

Element fire = Element
(
    element_words
    {
      .names = element_set_names({"fire", "flame", "blaze", "ember"}),
      .adjectives = element_set_adjectives({"fiery", "flaming", "burning", "flaring"}) 
    }
);

Element holiness = Element
(
    element_words
    {
      .names = element_set_names({"holiness"}),
      .adjectives = element_set_adjectives({"hole"}) 
    }
);


Element ice = Element
(
    element_words
    {
      .names = element_set_names({"ice", "frost"}),
      .adjectives = element_set_adjectives({"icy", "frozen", "frosty", "frosted"}) 
    }
);

Element metal  = Element
(
    element_words
    {
      .names = element_set_names({"metal", "steel"}),
      .adjectives = element_set_adjectives({"metallic"}) 
    }
);

Element mutation  = Element
(
    element_words
    {
      .names = element_set_names({"mutation"}),
      .adjectives = element_set_adjectives({"mutant", "mutated"}) 
    }
);

Element normal  = Element
(
    element_words
    {
      .names = element_set_names({"normal"}),
      .adjectives = element_set_adjectives({""}) 
    }
);


Element plant  = Element
(
    element_words
    {
      .names = element_set_names({"plant"}),
      .adjectives = element_set_adjectives({"grassy", "mossy", "leafy"}) 
    }
);

Element poison  = Element
(
    element_words
    {
      .names = element_set_names({"poison", "toxic"}),
      .adjectives = element_set_adjectives({"poisonous"}) 
    }
);

Element radiation  = Element
(
    element_words
    {
      .names = element_set_names({"radiation", "radioactivity"}),
      .adjectives = element_set_adjectives({"nuclear", "atomic", "radioactive"}) 
    }
);

Element shadow  = Element
(
    element_words
    {
      .names = element_set_names({"shadow", "darkness"}),
      .adjectives = element_set_adjectives({"shadowy", "dark"}) 
    }
);

Element undead = Element
(
    element_words
    {
      .names = element_set_names({"undead", "zombie"}),
      .adjectives = element_set_adjectives({"zombified", "rotting"}) 
    }
);

Element water = Element
(
    element_words
    {
      .names = element_set_names({"water", "sea"}),
      .adjectives = element_set_adjectives({"watery", "wet"}) 
    }
);

std::map<std::string, Element*> element_map =
{
    {"alien", &alien},
    {"air", &air},
    {"chaos", &chaos},
    {"earth", &earth},
    {"electricity", &electricity},
    {"evil", &evil},
    {"fire", &fire},
    {"holiness", &holiness},
    {"ice", &ice},
    {"metal", &metal},
    {"mutation", &mutation},
    {"normal", &normal},
    {"plant", &plant},
    {"poison", &poison},
    {"radiation", &radiation},
    {"undead", &undead},
    {"water", &water}
};

vec_str element_map_keys(std::map<std::string, Element*> element_map) {
    vec_str retval;

    for (auto el : element_map)
    {
      retval.emplace_back(el.first);
    }

    return retval;
}

File: python/backup/element/element.py
class Element:
    def __init__(self, name: str = "DEFAULT_ELEMENT_NAME", element_type: str = "DEFAULT_ELEMENT_TYPE", weak = [], strong = []):
        self.name = name
        self.element_type = element_type
        self.weak = weak
        self.strong = strong

    def info(self):
        print("+|==|+ ELEMENT INFO +|==|+")
        print("NAME:", self.name)
        print("ELEMENT_TYPE:", self.element_type)
        print()

DEFAULT_ELEMENT = Element(
)

alien = Element(
    name = "alien",
    element_type = "physical",
    weak = ["alien", "radiation", "undead", "water"],
    strong = ["air", "chaos", "earth", "electricity", "fire", "normal"],
)

air = Element(
    name = "air",
    element_type = "special",
    weak = ["alien", "air", "earth", "evil", "radiation", "plant"],
    strong = ["electricity", "fire", "ice", "mutation"],
)

chaos = Element(
    name = "chaos",
    element_type = "physical",
    weak = ["air", "earth", "fire", "plant"],
    strong = ["alien", "chaos", "mutation", "radiation", "undead", "water"],
)


earth = Element(
    name = "earth",
    element_type = "physical",
    weak = ["earth", "metal", "plant", "water"],
    strong = ["evil", "fire", "undead"],
)

electricity = Element(
    name = "electricity",
    element_type = "special",
    weak = ["alien", "air", "earth", "electricity", "plant"],
    strong = ["holiness", "metal", "water", "undead"],
)

evil = Element(
    name = "evil",
    element_type = "physical",
    weak = ["earth", "evil", "radiation", "undead"],
    strong = ["holiness", "normal", "plant", "ice"],
)

fire = Element(
    name = "fire",
    element_type = "special",
    weak = ["air", "earth", "evil", "fire", "water"],
    strong = ["alien", "ice", "metal", "plant", "undead"]
)

holiness = Element(
    name = "holiness",
    element_type = "physical",
    weak = ["alien", "chaos", "holiness"],
    strong = ["earth", "evil", "fire", "ice", "mutation", "radiation", "poison", "undead"],
)

ice = Element(
    name = "ice",
    element_type = "special",
    weak = ["chaos", "fire", "ice"],
    strong = ["earth", "holiness", "plant", "undead", "water"]
)

metal = Element(
    name = "metal",
    element_type = "physical",
    weak = ["metal", "water"],
    strong = ["air", "poison"],
)

mutation = Element(
    name = "mutation",
    element_type = "physical",
    weak = ["holiness", "ice", "mutation", "water"],
    strong = ["alien", "chaos", "evil", "radiation", "plant"],
)

normal = Element(
    name = "normal",
    element_type = "physical",
    weak = ["alien", "chaos", "undead"],
    strong = ["chaos", "earth", "evil", "mutation"]
)

plant = Element(
    name = "plant",
    element_type = "special",
    weak = ["chaos", "evil", "fire", "ice", "radiation", "plant"],
    strong = ["alien", "air", "earth", "holiness", "water"],
)

poison = Element(
    name = "poison",
    element_type = "physical",
    weak = ["metal", "mutation", "radiation", "poison", "undead"],
    strong = ["alien", "chaos", "holiness", "plant", "water"],
)

radiation = Element(
    name = "radiation",
    element_type = "special",
    weak = ["electricity", "fire", "undead"],
    strong = ["alien", "chaos", "radiation", "plant"],
)

undead = Element(
    name = "undead",
    element_type = "physical",
    weak = ["air", "earth", "fire", "metal", "undead", "water"],
    strong = ["alien", "chaos", "electricity", "holiness", "ice", "normal", "poison"]
)

water = Element(
    name = "water",
    element_type = "special",
    weak = ["chaos", "electricity", "ice", "mututation", "plant", "poison", "water"],
    strong = ["alien", "evil", "fire", "metal", "radiation"]
)

elements = {
    "DEFAULT_ELEMENT": DEFAULT_ELEMENT,
    "alien": alien,
    "air": air,
    "chaos": chaos,
    "earth": earth,
    "electricity": electricity,
    "evil": evil,
    "fire": fire,
    "holiness": holiness,
    "ice": ice,
    "metal": metal,
    "mutation": mutation,
    "normal": normal,
    "plant": plant,
    "poison": poison,
    "radiation": radiation,
    "undead": undead,
    "water": water
}
